usr/src/uts/common/fs/zfs/vdev_indirect.c

```diff
@@ -1492,12 +1496,49 @@ vdev_indirect_reconstruct_io_done(zio_t *zio)
 {
 	indirect_vsd_t *iv = zio->io_vsd;
 	uint64_t attempts = 0;
-	uint64_t attempts_max = 1ULL << zfs_reconstruct_indirect_segments_max;
-	int segments = 0;
+	uint64_t attempts_max = UINT64_MAX;
+	uint64_t combinations = 1;
+
+	if (zfs_reconstruct_indirect_combinations_max > 0)
+		attempts_max = zfs_reconstruct_indirect_combinations_max;
 
+	/*
+	 * Discard duplicate copies of split segments to minimize the
+	 * number of unique combinations when attempting reconstruction.
+	 */
 	for (indirect_split_t *is = list_head(&iv->iv_splits);
-	    is != NULL; is = list_next(&iv->iv_splits, is))
-		segments++;
+	    is != NULL; is = list_next(&iv->iv_splits, is)) {
+		uint64_t is_copies = 0;
+
+		for (int i = 0; i < is->is_children; i++) {
```

Comment:
I have a question which may lead a small optimization although I may be completely wrong.

In this loop we are going through each is_child:
If it has no data we skip it. If it has data we are looking at the rest of the is_child to see if any of them has the same data as our original is_child and mark them as duplicate, and finally we increment is_copies which we will use to get the number of combinations.

The above makes sense with the rest of the logic in this function but to me it sounds that what we call "accurate combinations" is not exactly accurate as the combinations themselves are not unique.

The variable `is_copies` (that is later used to calculate `combinations`) tells us how many of the is_children vdevs have their `ic_data` field populated (e.g. not NULL). So that got me wondering, what if:
[1] we changed the name of `is_copies` to something like `is_split_versions`
[2] in this loop instead of just skipping when `ic_data == NULL` we also skip when `ic_duplicate != -1` (meaning we visited this `ic_child` before [in the inner loop] and we know that it has not a unique version of `ic_data`)
[3] Later the outer loop (right after the inner loop is done) we increment `is_split_versions` and we use it exactly like `is_copies`.

This should give us the number of unique combinations. Now I know that this won't work exactly with the rest of the logic in this function but what if the struct `indirect_split_t` had a field `list_t is_unique_versions` (or some similar name) and every time `is_split_versions` is incremented in this loop we add the current `is_child` to this list.

Then later in this function we can use this new list without having to care, do checks, and skip duplicates because they are never visited. Thoughts?

Reply:
@sdimitro thanks for taking the time to review this.

> "accurate combinations" is not exactly accurate as the combinations themselves are not unique.

Good catch.  This appears to be a issue which snuck in when the patch was reworked.  The intent was to calculate unique copies and the original patch did this by immediately freeing any dulicate `is->is_child[i].ic_data` copies and setting ` is->is_child[i].ic_data = NULL`.

Subsequently the mechanism was reworked and `is->is_child[j].ic_duplicate` was added instead because `vdev_indirect_repair()` relied on having non-NULL `abd_t`'s for repair.  In the process it looks like the loop logic wasn't entirely updated.

What you're proposing sounds like a reasonable way to fix this.  I'll see about implementing the proposed changes in ZoL for @ahrens to add to this PR for review.

Reply:
@behlendorf let me know when you have the ZoL PR for this, and I'll pull it in here.